code/modules/xenoarchaeology/traits/activators.dm

```diff
@@ -0,0 +1,664 @@
+
+	setup_generic_item_hint()
+	addtimer(CALLBACK(src, PROC_REF(do_sonar)), 5 SECONDS)
+
+/datum/xenoartifact_trait/minor/signaller/Destroy(force, ...)
+	SSradio.remove_object(src, FREQ_SIGNALER)
+	QDEL_NULL(signal)
+	return ..()
+
+/datum/xenoartifact_trait/activator/signal/generate_trait_appearance(atom/movable/target)
+	. = ..()
+	if(!ismovable(target))
+		return
+	//Build particle holder
+	particle_holder = new(parent?.parent)
+	particle_holder.add_emitter(/obj/emitter/sonar, "sonar", 9)
+	//Layer onto parent
+	target.vis_contents += particle_holder
+
+/datum/xenoartifact_trait/activator/signal/cut_trait_appearance(atom/movable/target)
+	. = ..()
+	if(!ismovable(target))
+		return
+	target.vis_contents -= particle_holder
+	QDEL_NULL(particle_holder)
+
+/datum/xenoartifact_trait/activator/signal/do_hint(mob/user, atom/item)
+	if(istype(item, /obj/item/analyzer))
+		to_chat(user, "<span class='warning'>[item] detects an input frequency & code of [FREQ_SIGNALER]-[code]!</span>")
+		return ..()
+
+/datum/xenoartifact_trait/activator/signal/get_dictionary_hint()
+	. = ..()
+	return list(XENOA_TRAIT_HINT_TRIGGER("signaller assembly"),
+	XENOA_TRAIT_HINT_DETECT("analyzer, which will also reveal its trigger code & frequency"),
+	XENOA_TRAIT_HINT_RANDOMISED, list("icon" = "exclamation", "desc" = "This trait will activate on the nearest living target."),
+	XENOA_TRAIT_HINT_APPEARANCE("This trait will make radar particles appear around the artifact."),
+	XENOA_TRAIT_HINT_SOUND("sonar pings"))
+
+/datum/xenoartifact_trait/activator/signal/proc/receive_signal(datum/signal/signal)
+	if(!signal)
+		return
+	if(signal.data["code"] != code)
+		return
+	for(var/atom/target in oview(parent.target_range, get_turf(parent?.parent)))
+		//Only add mobs
+		if(!ismob(target))
+			continue
+		trigger_artifact(target, XENOA_ACTIVATION_CONTACT)
+		break
+	if(!length(parent.targets))
+		parent.trigger()
+
+/datum/xenoartifact_trait/activator/signal/proc/do_sonar(repeat = TRUE)
+	if(QDELETED(src))
+		return
+	var/atom/A = parent.parent
+	if(isturf(A.loc))
+		playsound(get_turf(parent?.parent), 'sound/effects/ping.ogg', 60, TRUE)
+	var/rand_time = rand(5, 15) SECONDS
+	addtimer(CALLBACK(src, PROC_REF(do_sonar)), rand_time / (isturf(A.loc) ? 2 : 1))
+
+/*
+	ABSTRACT
+	Item key
+	This trait activates when an item key is used on it
+*/
+/datum/xenoartifact_trait/activator/item_key
+	flags = XENOA_HIDE_TRAIT
+	///What item type activates us?
+	var/obj/item/key_item
+	///Is the key item a strict type?
+	var/is_strict = FALSE
+
+/datum/xenoartifact_trait/activator/item_key/New(atom/_parent)
+	. = ..()
+	if(!parent?.parent)
+		return
+	RegisterSignal(parent?.parent, COMSIG_PARENT_ATTACKBY, TYPE_PROC_REF(/datum/xenoartifact_trait/activator, translation_type_b))
+
+/datum/xenoartifact_trait/activator/item_key/translation_type_b(datum/source, atom/item, atom/target)
+	return (is_strict ? item?.type == key_item.type : istype(item, key_item)) && !parent.use_cooldown_timer
```

Comment:
The readability of this code is extremely poor, it is almost impossible to tell what this does